File: traceharbor-resource-canary/traceharbor-resource-canary-android/src/main/java/io/traceharbor/resource/processor/BaseLeakProcessor.java
package io.traceharbor.resource.processor;

import android.os.Build;

import io.traceharbor.TraceHarbor;
import io.traceharbor.report.Issue;
import io.traceharbor.resource.CanaryWorkerService;
import io.traceharbor.resource.ResourcePlugin;
import io.traceharbor.resource.analyzer.ActivityLeakAnalyzer;
import io.traceharbor.resource.analyzer.model.ActivityLeakResult;
import io.traceharbor.resource.analyzer.model.AndroidExcludedRefs;
import io.traceharbor.resource.analyzer.model.DestroyedActivityInfo;
import io.traceharbor.resource.analyzer.model.ExcludedRefs;
import io.traceharbor.resource.analyzer.model.HeapDump;
import io.traceharbor.resource.analyzer.model.HeapSnapshot;
import io.traceharbor.resource.config.ResourceConfig;
import io.traceharbor.resource.config.SharePluginInfo;
import io.traceharbor.resource.dumper.AndroidHeapDumper;
import io.traceharbor.resource.watcher.ActivityRefWatcher;
import io.traceharbor.util.TraceHarborLog;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.IOException;

/**
 * Created by Yves on 2021/2/25
 */
public abstract class BaseLeakProcessor {
    private static final String TAG = "TraceHarbor.LeakProcessor.Base";

    private final ActivityRefWatcher mWatcher;

    private AndroidHeapDumper                 mHeapDumper;
    private AndroidHeapDumper.HeapDumpHandler mHeapDumpHandler;

    public BaseLeakProcessor(ActivityRefWatcher watcher) {
        mWatcher = watcher;
    }

    public abstract boolean process(DestroyedActivityInfo destroyedActivityInfo);

    @Deprecated
    public AndroidHeapDumper getHeapDumper() {
        if (mHeapDumper == null) {
            mHeapDumper = new AndroidHeapDumper(mWatcher.getContext());
        }
        return mHeapDumper;
    }

    protected AndroidHeapDumper.HeapDumpHandler getHeapDumpHandler() {
        if (mHeapDumpHandler == null) {
            mHeapDumpHandler = new AndroidHeapDumper.HeapDumpHandler() {
                @Override
                public void process(HeapDump result) {
                    CanaryWorkerService.shrinkHprofAndReport(mWatcher.getContext(), result);
                }
            };
        }

        return mHeapDumpHandler;
    }

    public ActivityRefWatcher getWatcher() {
        return mWatcher;
    }

    public void onDestroy() {
    }

    private static volatile boolean mAnalyzing = false;

    public static boolean isAnalyzing() {
        return mAnalyzing;
    }

    private static void setAnalyzing(boolean analyzing) {
        mAnalyzing = analyzing;
    }

    protected ActivityLeakResult analyze(File hprofFile, String referenceKey) {
        setAnalyzing(true);
        final HeapSnapshot heapSnapshot;
        ActivityLeakResult result;
        String manufacture = TraceHarbor.with().getPluginByClass(ResourcePlugin.class).getConfig().getManufacture();
        final ExcludedRefs excludedRefs = AndroidExcludedRefs.createAppDefaults(Build.VERSION.SDK_INT, manufacture).build();
        try {
            heapSnapshot = new HeapSnapshot(hprofFile);
            result = new ActivityLeakAnalyzer(referenceKey, excludedRefs).analyze(heapSnapshot);
        } catch (IOException e) {
            result = ActivityLeakResult.failure(e, 0);
        }
        getWatcher().triggerGc();
        setAnalyzing(false);
        return result;
    }

    final protected void publishIssue(int issueType, ResourceConfig.DumpMode dumpMode, String activity, String refKey, String detail, String cost) {
        publishIssue(issueType, dumpMode, activity, refKey, detail, cost, 0);
    }

    final protected void publishIssue(int issueType, ResourceConfig.DumpMode dumpMode, String activity, String refKey, String detail, String cost, int retryCount) {
        publishIssue(issueType, dumpMode, activity, refKey, detail, cost, retryCount, null);
    }

    final protected void publishIssue(int issueType, ResourceConfig.DumpMode dumpMode, String activity, String refKey, String detail, String cost, int retryCount, String hprofPath) {
        Issue issue = new Issue(issueType);
        JSONObject content = new JSONObject();
        try {
            content.put(SharePluginInfo.ISSUE_DUMP_MODE, dumpMode.name());
            content.put(SharePluginInfo.ISSUE_ACTIVITY_NAME, activity);
            content.put(SharePluginInfo.ISSUE_REF_KEY, refKey);
            content.put(SharePluginInfo.ISSUE_LEAK_DETAIL, detail);
            content.put(SharePluginInfo.ISSUE_COST_MILLIS, cost);
            content.put(SharePluginInfo.ISSUE_RETRY_COUNT, retryCount);
            content.put(SharePluginInfo.ISSUE_HPROF_PATH, hprofPath);
        } catch (JSONException jsonException) {
            TraceHarborLog.printErrStackTrace(TAG, jsonException, "");
        }
        issue.setContent(content);
        getWatcher().getResourcePlugin().onDetectIssue(issue);
    }
}

File: traceharbor-resource-canary/traceharbor-resource-canary-android/src/main/java/com/kernelflux/traceharbor/resource/processor/BaseLeakProcessor.kt
package com.kernelflux.traceharbor.resource.processor

import android.os.Build
import com.kernelflux.traceharbor.TraceHarbor
import com.kernelflux.traceharbor.report.Issue
import com.kernelflux.traceharbor.resource.CanaryWorkerService
import com.kernelflux.traceharbor.resource.ResourcePlugin
import com.kernelflux.traceharbor.resource.analyzer.ActivityLeakAnalyzer
import com.kernelflux.traceharbor.resource.analyzer.model.ActivityLeakResult
import com.kernelflux.traceharbor.resource.analyzer.model.AndroidExcludedRefs
import com.kernelflux.traceharbor.resource.analyzer.model.DestroyedActivityInfo
import com.kernelflux.traceharbor.resource.analyzer.model.ExcludedRefs
import com.kernelflux.traceharbor.resource.analyzer.model.HeapDump
import com.kernelflux.traceharbor.resource.analyzer.model.HeapSnapshot
import com.kernelflux.traceharbor.resource.config.ResourceConfig
import com.kernelflux.traceharbor.resource.config.SharePluginInfo
import com.kernelflux.traceharbor.resource.dumper.AndroidHeapDumper
import com.kernelflux.traceharbor.resource.watcher.ActivityRefWatcher
import com.kernelflux.traceharbor.util.TraceHarborLog
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.io.IOException
import kotlin.jvm.JvmName

abstract class BaseLeakProcessor(
    @JvmField protected val mWatcher: ActivityRefWatcher,
) {
    private var mHeapDumper: AndroidHeapDumper? = null
    private var mHeapDumpHandler: AndroidHeapDumper.HeapDumpHandler? = null

    abstract fun process(destroyedActivityInfo: DestroyedActivityInfo): Boolean

    @Deprecated("Kept for Java compatibility")
    fun getHeapDumper(): AndroidHeapDumper {
        if (mHeapDumper == null) {
            mHeapDumper = AndroidHeapDumper(mWatcher.context)
        }
        return mHeapDumper!!
    }

    protected fun getHeapDumpHandler(): AndroidHeapDumper.HeapDumpHandler {
        if (mHeapDumpHandler == null) {
            mHeapDumpHandler =
                object : AndroidHeapDumper.HeapDumpHandler {
                    override fun process(result: HeapDump) {
                        CanaryWorkerService.shrinkHprofAndReport(mWatcher.context, result)
                    }
                }
        }
        return mHeapDumpHandler!!
    }

    fun getWatcher(): ActivityRefWatcher = mWatcher

    @get:JvmName("getWatcherCompat")
    protected val watcher: ActivityRefWatcher
        get() = mWatcher

    @get:JvmName("getHeapDumperCompat")
    protected val heapDumper: AndroidHeapDumper
        @Suppress("DEPRECATION")
        get() = getHeapDumper()

    @get:JvmName("getHeapDumpHandlerCompat")
    protected val heapDumpHandler: AndroidHeapDumper.HeapDumpHandler
        get() = getHeapDumpHandler()

    open fun onDestroy() {}

    protected fun analyze(hprofFile: File, referenceKey: String): ActivityLeakResult {
        setAnalyzing(true)
        var result: ActivityLeakResult
        val plugin = TraceHarbor.with().getPluginByClass(ResourcePlugin::class.java)
        val manufacture =
            plugin?.getConfig()?.getManufacture() ?: ""
        val excludedRefs: ExcludedRefs =
            AndroidExcludedRefs.createAppDefaults(Build.VERSION.SDK_INT, manufacture).build()
        try {
            val heapSnapshot = HeapSnapshot(hprofFile)
            result = ActivityLeakAnalyzer(referenceKey, excludedRefs).analyze(heapSnapshot)
        } catch (e: IOException) {
            result = ActivityLeakResult.failure(e, 0)
        }
        mWatcher.triggerGc()
        setAnalyzing(false)
        return result
    }

    protected fun publishIssue(
        issueType: Int,
        dumpMode: ResourceConfig.DumpMode,
        activity: String,
        refKey: String,
        detail: String,
        cost: String,
    ) {
        publishIssue(issueType, dumpMode, activity, refKey, detail, cost, 0)
    }

    protected fun publishIssue(
        issueType: Int,
        dumpMode: ResourceConfig.DumpMode,
        activity: String,
        refKey: String,
        detail: String,
        cost: String,
        retryCount: Int,
    ) {
        publishIssue(issueType, dumpMode, activity, refKey, detail, cost, retryCount, null)
    }

    protected fun publishIssue(
        issueType: Int,
        dumpMode: ResourceConfig.DumpMode,
        activity: String,
        refKey: String,
        detail: String,
        cost: String,
        retryCount: Int,
        hprofPath: String?,
    ) {
        val issue = Issue(issueType)
        val content = JSONObject()
        try {
            content.put(SharePluginInfo.ISSUE_DUMP_MODE, dumpMode.name)
            content.put(SharePluginInfo.ISSUE_ACTIVITY_NAME, activity)
            content.put(SharePluginInfo.ISSUE_REF_KEY, refKey)
            content.put(SharePluginInfo.ISSUE_LEAK_DETAIL, detail)
            content.put(SharePluginInfo.ISSUE_COST_MILLIS, cost)
            content.put(SharePluginInfo.ISSUE_RETRY_COUNT, retryCount)
            content.put(SharePluginInfo.ISSUE_HPROF_PATH, hprofPath)
        } catch (jsonException: JSONException) {
            TraceHarborLog.printErrStackTrace(TAG, jsonException, "")
        }
        issue.content = content
        mWatcher.getResourcePlugin().onDetectIssue(issue)
    }

    companion object {
        private const val TAG = "TraceHarbor.LeakProcessor.Base"

        @Volatile
        private var mAnalyzing: Boolean = false

        @JvmStatic
        fun isAnalyzing(): Boolean = mAnalyzing

        private fun setAnalyzing(analyzing: Boolean) {
            mAnalyzing = analyzing
        }
    }
}

